File: office-scripts/PublishEngine.ts
// ============================================================================
// FIELD FORCE OPTIMIZER V11 - PUBLISH ENGINE
// ============================================================================
// Deployable Office Script. Run manually, as an explicit action, after
// reviewing/adjusting the Draft plan in MANAGER_PLAN and before sending it to
// technicians. This is the "Publish" step from docs/BUSINESS_RULES.md
// section 11 (Draft -> Review -> Published -> Active -> Closed).
//
// WHAT IT DOES:
//   - Finds the single EARLIEST week that is still Draft (has rows in
//     MANAGER_PLAN but no Published/Active/Closed row in PLAN_LIFECYCLE).
//     Only one week is published per run, matching the real weekly ritual
//     (you publish the upcoming week, not the whole multi-week horizon at
//     once) and the "only the nearest week becomes binding" rolling-horizon
//     design agreed earlier.
//   - Copies that week's MANAGER_PLAN rows verbatim into
//     MANAGER_PLAN_PUBLISHED (append-only, immutable snapshot).
//   - Sets that week's PLAN_LIFECYCLE status to Published, records
//     publishedAt.
//
// WHAT IT DELIBERATELY DOES NOT DO:
//   - It does not touch MANAGER_PLAN itself - Planning Engine already
//     refuses to regenerate a locked week once PLAN_LIFECYCLE says
//     Published, so simply flipping the status here is sufficient to lock it
//     going forward.
//   - It does not decide when Published becomes Active or Closed - that is
//     mechanical/derived and handled by ComplianceEngine.ts on every run,
//     never here.
//   - It does not pick a technician-by-technician subset - a week is
//     published as a whole (matches "tyden je zavazny", not per-technician).
// ============================================================================

function main(workbook: ExcelScript.Workbook) {
  function readTable(sheetName: string): (string | number | boolean)[][] {
    const ws = workbook.getWorksheet(sheetName);
    const range = ws.getUsedRange();
    return range ? range.getValues() : [];
  }

  const managerPlan = readTable("MANAGER_PLAN");
  const planLifecycle = readTable("PLAN_LIFECYCLE");
  const control = readTable("CONTROL");

  function setting(name: string, fallback: number): number {
    for (let i = 1; i < control.length; i++) {
      if (String(control[i][0]).toUpperCase().trim() == name.toUpperCase()) {
        const v = Number(control[i][1]);
        return isNaN(v) ? fallback : v;
      }
    }
    return fallback;
  }
  const YEAR = setting("YEAR", new Date().getFullYear());

  if (managerPlan.length < 2) {
    console.log("Publish Engine: MANAGER_PLAN is empty - run Planning Engine first.");
    return;
  }

  // Weeks already Published/Active/Closed - never republish.
  let lockedWeeks = new Set<number>();
  let plHeaders: string[] = [];
  let plIdx: (name: string) => number = () => -1;
  if (planLifecycle.length >= 2) {
    plHeaders = (planLifecycle[0] as string[]).map((h) => String(h));
    plIdx = (name: string) => plHeaders.indexOf(name);
    for (let i = 1; i < planLifecycle.length; i++) {
      const row = planLifecycle[i];
      if (Number(row[plIdx("year")]) != YEAR) {
        continue;
      }
      const status = String(row[plIdx("status")]);
      if (status == "Published" || status == "Active" || status == "Closed") {
        lockedWeeks.add(Number(row[plIdx("week")]));
      }
    }
  }

  // Earliest Draft week present in MANAGER_PLAN.
  let draftWeeks = new Set<number>();
  for (let i = 1; i < managerPlan.length; i++) {
    const week = Number(managerPlan[i][0]);
    if (week && !lockedWeeks.has(week)) {
      draftWeeks.add(week);
    }
  }
  if (draftWeeks.size == 0) {
    console.log("Publish Engine: no Draft week found to publish (everything is already locked, or MANAGER_PLAN is empty).");
    return;
  }
  const weekToPublish = Math.min(...draftWeeks);

  const rowsToPublish = managerPlan
    .slice(1)
    .filter((row) => Number(row[0]) == weekToPublish);

  const now = new Date().toISOString();
  const publishedRows = rowsToPublish.map((row) => [...row, now] as (string | number)[]);

  const publishedWs = workbook.getWorksheet("MANAGER_PLAN_PUBLISHED");
  const existingPublished = publishedWs.getUsedRange();
  const startRow = existingPublished ? existingPublished.getRowCount() : 1;
  publishedWs.getRangeByIndexes(startRow, 0, publishedRows.length, 18).setValues(publishedRows);

  // Update or insert the PLAN_LIFECYCLE row for this week.
  const plWs = workbook.getWorksheet("PLAN_LIFECYCLE");
  let existingRowIndex = -1;
  for (let i = 1; i < planLifecycle.length; i++) {
    if (Number(planLifecycle[i][plIdx("year")]) == YEAR && Number(planLifecycle[i][plIdx("week")]) == weekToPublish) {
      existingRowIndex = i;
      break;
    }
  }
  if (existingRowIndex >= 0) {
    plWs.getRangeByIndexes(existingRowIndex, 2, 1, 2).setValues([["Published", now]]);
  } else {
    const startPlRow = planLifecycle.length > 0 ? planLifecycle.length : 1;
    plWs.getRangeByIndexes(startPlRow, 0, 1, 5).setValues([[YEAR, weekToPublish, "Published", now, ""]]);
  }

  console.log(
    "Publish Engine: week " + weekToPublish + "/" + YEAR + " published (" +
      publishedRows.length + " visits snapshotted to MANAGER_PLAN_PUBLISHED). " +
      "This week is now locked - Planning Engine will not regenerate it."
  );
}

File: desktop_client/engines/publish_engine.py
"""
Python port of office-scripts/PublishEngine.ts's main(). Line-for-line
translation - see import_engine.py's module docstring for the duplication
rationale and mock_workbook.py for why the class-based Range API is kept.
"""
from __future__ import annotations

import datetime

from .core_logic import iso_now
from .js_compat import at as _at
from .mock_workbook import MockWorkbook


def run(workbook: MockWorkbook) -> str:
    def read_table(sheet_name: str) -> list[list]:
        ws = workbook.get_worksheet(sheet_name)
        rng = ws.get_used_range()
        return rng.get_values() if rng else []

    manager_plan = read_table("MANAGER_PLAN")
    plan_lifecycle = read_table("PLAN_LIFECYCLE")
    control = read_table("CONTROL")

    def setting(name: str, fallback: float) -> float:
        for i in range(1, len(control)):
            if str(control[i][0]).upper().strip() == name.upper():
                try:
                    return float(control[i][1])
                except (TypeError, ValueError):
                    return fallback
        return fallback

    year = int(setting("YEAR", datetime.date.today().year))

    if len(manager_plan) < 2:
        return "Publish Engine: MANAGER_PLAN is empty - run Planning Engine first."

    locked_weeks: set[int] = set()
    pl_headers: list[str] = [str(h) for h in plan_lifecycle[0]] if len(plan_lifecycle) >= 2 else []

    def pl_idx(name: str) -> int:
        return pl_headers.index(name) if name in pl_headers else -1

    if len(plan_lifecycle) >= 2:
        for i in range(1, len(plan_lifecycle)):
            row = plan_lifecycle[i]
            if int(float(_at(row, pl_idx("year")) or 0)) != year:
                continue
            status = str(_at(row, pl_idx("status")))
            if status in ("Published", "Active", "Closed"):
                locked_weeks.add(int(float(_at(row, pl_idx("week")) or 0)))

    draft_weeks: set[int] = set()
    for i in range(1, len(manager_plan)):
        week_val = manager_plan[i][0]
        try:
            week = int(float(week_val)) if week_val not in (None, "") else 0
        except (TypeError, ValueError):
            week = 0
        if week and week not in locked_weeks:
            draft_weeks.add(week)

    if len(draft_weeks) == 0:
        return "Publish Engine: no Draft week found to publish (everything is already locked, or MANAGER_PLAN is empty)."

    week_to_publish = min(draft_weeks)

    rows_to_publish = [row for row in manager_plan[1:] if int(float(row[0])) == week_to_publish]

    now = iso_now()
    published_rows = [list(row) + [now] for row in rows_to_publish]

    published_ws = workbook.get_worksheet("MANAGER_PLAN_PUBLISHED")
    existing_published = published_ws.get_used_range()
    start_row = existing_published.get_row_count() if existing_published else 1
    published_ws.get_range_by_indexes(start_row, 0, len(published_rows), 18).set_values(published_rows)

    pl_ws = workbook.get_worksheet("PLAN_LIFECYCLE")
    existing_row_index = -1
    for i in range(1, len(plan_lifecycle)):
        row = plan_lifecycle[i]
        if int(float(_at(row, pl_idx("year")) or 0)) == year and int(float(_at(row, pl_idx("week")) or 0)) == week_to_publish:
            existing_row_index = i
            break
    if existing_row_index >= 0:
        pl_ws.get_range_by_indexes(existing_row_index, 2, 1, 2).set_values([["Published", now]])
    else:
        start_pl_row = len(plan_lifecycle) if len(plan_lifecycle) > 0 else 1
        pl_ws.get_range_by_indexes(start_pl_row, 0, 1, 5).set_values([[year, week_to_publish, "Published", now, ""]])

    return (
        f"Publish Engine: week {week_to_publish}/{year} published "
        f"({len(published_rows)} visits snapshotted to MANAGER_PLAN_PUBLISHED). "
        "This week is now locked - Planning Engine will not regenerate it."
    )
